Initial attempt at an API refactor

miguel-vila opened the pull request:
This is not intended to be a final PR and it's just to start a conversation. This is intended to address #52

A summary of some of the changes:

- Simplify `DogStatsDClient` and `StatsDClient` by making them depend on `Client` directly
- Make `Client` depend on an initialized queue
- Encapsulate queue creation and shutdown with constructors that return a `Managed`
- Comment out some constructors for the time being, could be replaced by mapping the `ZManaged`

Some open questions I still have:
- The `.orDie` in the `fiber.join` doesn't look good, not sure how that thread should be cleaned up(?)
- Not sure the shutdown order is right. Think `ClientTest` has been broken.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zio/zio-metrics?pullRequest=53) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/zio/zio-metrics?pullRequest=53) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zio/zio-metrics?pullRequest=53) it.</sub>

toxicafunk commented on statsd/src/main/scala/zio/metrics/statsd/StatsDClient.scala:
```diff
@@ -1,104 +1,94 @@
   }
 }
 
 object StatsDClient {
```
Companion object is apparently not needed anymore.

toxicafunk commented on statsd/src/main/scala/zio/metrics/dogstatsd/DogStatsDClient.scala:
```diff
@@ -160,29 +138,29 @@ class DogStatsDClient(
   }
 }
 
 object DogStatsDClient {
```
Not needed anymore it seems.

toxicafunk commented:
@miguel-vila the tests are fine, the problem is you need to also update the documentation, in particular `docs/essential/statsd.md`.

You can verufy by running `sbt docs/mdoc`

miguel-vila commented:
hello @toxicafunk 👋 , made some changes:

- 2db78b4: made some stuff private, added some start constructors for the specific clients, I hope those are useful.
- b84b533: updated the docs, this included changing some of the code and also rewording some stuff, I hope I covered everything and that it is clear, let me know how it looks to you and if you want to reword something or change anything go ahead.
